File: problems/leetcode/0000/56/56.py
# 56. Merge Intervals
# https://leetcode.com/problems/merge-intervals/

from typing import *

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        result = []

        for end, start in sorted([[e, s] for s, e in intervals]):
            while result and start <= result[-1][1]:
                start = min(start, result[-1][0])
                end = max(end, result[-1][1])
                result.pop()

            result.append([start, end])

        return result

File: problems/leetcode/0000/56/56.cpp
// 56. Merge Intervals
// https://leetcode.com/problems/merge-intervals/

using namespace std;

#include <vector>

class Solution {
public:
  vector<vector<int>> merge(vector<vector<int>> &intervals) {
    vector<vector<int>> answer;

    sort(intervals.begin(), intervals.end());
    answer.push_back(intervals[0]);

    int cur = 0;
    for (int i = 1; i < intervals.size(); i++) {
      if (answer[cur][1] < intervals[i][0]) {
        answer.push_back(intervals[i]);
        cur++;
      } else {
        answer[cur][1] = max(answer[cur][1], intervals[i][1]);
      }
    }

    return answer;
  }
};
